### ISSUE TITLE
Investigate how to allow event recovery without introducing duplicates

### ISSUE THREAD

**Issue by fblundun** [snowplow/snowplow#2684]
One possibility: we could have error JSONs for POSTs mention which specific event in the POST they concern. If this field isn't specified it means that they refer to the POST as a whole.

**Comment by alexanderdean** [comment]
I think that's a good idea Fred. Another suggestion is that we have some kind of intermediate representation of a failed event which is reported in place of the original raw line...

**Comment by fblundun** [comment]
It's an interesting trade-off - the intermediate representation makes it easier to do this kind of programmatic analysis of the bad rows, but also means we lose the ability to see exactly what went into the enrichment process. I guess we could expand the bad rows to contain both the raw line and the intermediate representation.

**Comment by alexanderdean** [comment]
Yes, we could have a structure like:

```
( raw payload, [ raw errors ] )
[ ( event 1, [ event 1 errors ] ),
  ( event 2, [ event 2 errors ] ),
  ( event 3, [ event 3 errors ] ),
]
```

The reason I quite like the intermediate form is that, over time, the bulk of our failures are going to relate to issues in "well-formed" events (such as schemas that can't be inferred), rather than issues with the raw payload itself. So going back to the raw payload each time feels a little wasteful...

**Comment by alexanderdean** [comment]
Or maybe we are thinking about the "bad rows" output wrongly, and we should simply be emitting well-formed Snowplow events for each event that fails on a 1:1 basis.

**Comment by yalisassoon** [comment]
Shouldn't we be separately solving deduplication? Then we can happily
reprocess away without worrying about accidentally reprocessing good data?

On Friday, May 20, 2016, Alexander Dean <notifications@github.com> wrote:

> Or maybe we are thinking about the "bad rows" output wrongly, and we
> should simply be emitting well-formed Snowplow events for each event that
> fails on a 1:1 basis.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/snowplow/snowplow/issues/2684#issuecomment-220559956>
>


-- 
Co-founder
Snowplow Analytics <http://snowplowanalytics.com/>
The Roma Building, 32-38 Scrutton Street, London EC2A 4RQ, United Kingdom
+44 (0)203 589 6116
+44 7841 954 117
@yalisassoon <https://twitter.com/yalisassoon>
<https://twitter.com/yalisassoon>

**Comment by alexanderdean** [comment]
Yes - we should be generally solving for the introduction of duplicates anywhere in the pipeline.

But thinking about the fact that our current bad rows output introduces a lot of duplicates needlessly when reprocessed (because its structure dates from when 1 payload :: 1 event) is a good starting point for coming up with a better bad rows format.

The duplicates arise because you can't tell which event in a POST payload a given failure applies to, which (regardless of the dupes) makes debugging harder. Another limitation of the current format is that it doesn't let you review the failed event in Elasticsearch. Debugging and recovery with the current bad rows format will get more challenging as we introduce a more complex enrichment DAG, and also schema inference...

There is a wider architectural issue too, which is that our `stderr`-style bad rows outputs (Kinesis or S3) don't compose. Our happy path composes well: we can route a unitary happy path all the way from collection through to storage targets. But on the failure path, we end up with all these individual silos of bad rows which need special-case treatment (like bad rows ingest into Elasticsearch).

I suspect that the answer down the line is to unify the "good" and "bad" output, so that there is just one output stream of events from a given component, containing a spectrum of:

1. Raw payloads which are corrupt
2. Events which have failed schema validation
3. Events which have failed enrichment
4. Events which pass all validation and enrichment but contain ?-?-? schemas which cannot (yet) be inferred
5. Events which pass all validation, enrichment and inference with warnings
6. Events which pass all validation, enrichment and inference with no warnings

So it's really a spectrum of (6) shades of grey, versus "good" and "bad". The nice thing about this approach is that the whole thing composes, because it's just events. And there's no need for special casing for bad rows versus good: if you can represent an enriched event in Avro or Redshift, you can represent an event which has failed enrichment or has warnings using the exact same tech...

**Comment by yalisassoon** [comment]
That's awesome - I hadn't thought about it in that way before...

On Fri, May 20, 2016 at 11:34 PM, Alexander Dean <notifications@github.com>
wrote:

> Yes - we should be generally solving for the introduction of duplicates
> anywhere in the pipeline.
>
> But thinking about the fact that our current bad rows output introduces a
> lot of duplicates needlessly when reprocessed (because its structure dates
> from when 1 payload :: 1 event) is a good starting point for coming up with
> a better bad rows format.
>
> The duplicates arise because you can't tell which event in a POST payload
> a given failure applies to, which (regardless of the dupes) makes debugging
> harder. Another limitation of the current format is that it doesn't let you
> review the failed event in Elasticsearch. Debugging and recovery with the
> current bad rows format will get more challenging as we introduce a more
> complex enrichment DAG, and also schema inference...
>
> There is a wider architectural issue too, which is that our stderr-style
> bad rows outputs (Kinesis or S3) don't compose. Our happy path composes
> well: we can route a unitary happy path all the way from collection through
> to storage targets. But on the failure path, we end up with all these
> individual silos of bad rows which need special-case treatment (like bad
> rows ingest into Elasticsearch).
>
> I suspect that the answer down the line is to unify the "good" and "bad"
> output, so that there is just one output stream of events from a given
> component, containing a spectrum of:
>
>    1. Raw payloads which are corrupt
>    2. Events which have failed schema validation
>    3. Events which have failed enrichment
>    4. Events which pass all validation and enrichment but contain ?-?-?
>    schemas which cannot (yet) be inferred
>    5. Events which pass all validation, enrichment and inference with
>    warnings
>    6. Events which pass all validation, enrichment and inference with no
>    warnings
>
> So it's really a spectrum of (6) shades of grey, versus "good" and "bad".
> The nice thing about this approach is that the whole thing composes,
> because it's just events. And there's no need for special casing for bad
> rows versus good: if you can represent an enriched event in Avro or
> Redshift, you can represent an event which has failed enrichment or has
> warnings using the exact same tech...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/snowplow/snowplow/issues/2684#issuecomment-220734389>
>



-- 
Co-founder
Snowplow Analytics <http://snowplowanalytics.com/>
The Roma Building, 32-38 Scrutton Street, London EC2A 4RQ, United Kingdom
+44 (0)203 589 6116
+44 7841 954 117
@yalisassoon <https://twitter.com/yalisassoon>
<https://twitter.com/yalisassoon>

**Comment by alexanderdean** [comment]
See also #351

**Comment by alexanderdean** [comment]
Let's move this into the new protocols repo.
